engine/src/main/java/org/terasology/rendering/world/WorldRendererImpl.java

```diff
@@ -618,45 +608,50 @@ private void renderSimpleBlendMaterials() {
         PerformanceMonitor.endActivity();
     }
 
-    private void renderChunk(RenderableChunk chunk, ChunkMesh.RenderPhase phase, Camera camera, ChunkRenderMode mode) {
-        if (chunk.hasMesh()) {
-            final Vector3f cameraPosition = camera.getPosition();
-            final Vector3f chunkPosition = chunk.getPosition().toVector3f();
-            final Vector3f chunkPositionRelativeToCamera =
-                    new Vector3f(chunkPosition.x * ChunkConstants.SIZE_X - cameraPosition.x,
-                            chunkPosition.y * ChunkConstants.SIZE_Y - cameraPosition.y,
-                            chunkPosition.z * ChunkConstants.SIZE_Z - cameraPosition.z);
-
-            if (mode == ChunkRenderMode.DEFAULT || mode == ChunkRenderMode.REFLECTION) {
-                if (phase == ChunkMesh.RenderPhase.REFRACTIVE) {
-                    chunkShader.activateFeature(ShaderProgramFeature.FEATURE_REFRACTIVE_PASS);
-                } else if (phase == ChunkMesh.RenderPhase.ALPHA_REJECT) {
-                    chunkShader.activateFeature(ShaderProgramFeature.FEATURE_ALPHA_REJECT);
-                }
+    private void renderChunk(PriorityQueue<RenderableChunk> chunks, ChunkMesh.RenderPhase phase, Camera camera, ChunkRenderMode mode) {
```

Comment:
The diff does not do these changes justice.  Basically I moved up all the stuff that is the same for all the chunks (camera position, logic about render phase, and common uniforms), and then made sure all uniform setting used the "currentOnly = true".

Most importantly moved up the ```chunkShader.activateFeature(...)``` because that causes the shader to become dirty and re-enable the shader with openGL.